diff --git a/src/open_clip/__init__.py b/src/open_clip/__init__.py
index d0419b4..49b5911 100644
--- a/src/open_clip/__init__.py
+++ b/src/open_clip/__init__.py
@@ -16,3 +16,4 @@ from .tokenizer import SimpleTokenizer, tokenize, decode
 from .transform import image_transform, AugmentationCfg
 from .zero_shot_classifier import build_zero_shot_classifier, build_zero_shot_classifier_legacy
 from .zero_shot_metadata import OPENAI_IMAGENET_TEMPLATES, SIMPLE_IMAGENET_TEMPLATES, IMAGENET_CLASSNAMES
+from .mobileclip2 import fastvit_mci3, fastvit_mci4
